File: src/conventions/definitions/system-config.ts
/**
 * Base bidding system identifiers and system-level bidding configuration.
 *
 * Captures HCP thresholds and ranges that vary between bidding systems
 * (SAYC, 2/1 Game Forcing, Acol, etc.) but are shared across all
 * conventions that plug into that system.
 *
 * Convention modules MUST receive a SystemConfig via dependency injection
 * (factory function parameter) — they must NOT hardcode these values.
 *
 * Convention-intrinsic values (e.g. "Stayman requires a 4-card major")
 * are NOT part of SystemConfig; they belong to the convention module.
 */

// ─── Known system IDs ───────────────────────────────────────

/** Standard American Yellow Card (SAYC). */
export const BASE_SYSTEM_SAYC = "sayc" as const;

/** Two-Over-One Game Forcing. */
export const BASE_SYSTEM_TWO_OVER_ONE = "two-over-one" as const;

/** Acol (standard weak-NT variant, UK tournament bridge). */
export const BASE_SYSTEM_ACOL = "acol" as const;

/** All known base-system identifiers. Extend this union as new systems are added. */
export type BaseSystemId =
  | typeof BASE_SYSTEM_SAYC
  | typeof BASE_SYSTEM_TWO_OVER_ONE
  | typeof BASE_SYSTEM_ACOL;

// ─── 1NT system parameters ──────────────────────────────────

export interface NtOpeningConfig {
  /** Minimum HCP for a 1NT opening (e.g. 15 in SAYC, 12 in Acol). */
  readonly minHcp: number;
  /** Maximum HCP for a 1NT opening (e.g. 17 in SAYC, 14 in Acol). */
  readonly maxHcp: number;
}

export interface ResponderThresholds {
  /** Minimum HCP for an invitational response (e.g. 8 in SAYC). */
  readonly inviteMin: number;
  /** Maximum HCP for an invitational response (e.g. 9 in SAYC). */
  readonly inviteMax: number;
  /** Minimum HCP for a game-forcing response (e.g. 10 in SAYC). */
  readonly gameMin: number;
  /** Minimum HCP for slam exploration (e.g. 15 in SAYC). */
  readonly slamMin: number;
}

export interface OpenerRebidThresholds {
  /** HCP at which opener is "not minimum" — typically ntOpening.minHcp + 1. */
  readonly notMinimum: number;
}

export interface InterferenceThresholds {
  /** Minimum HCP to redouble after opponent doubles (e.g. 10 in SAYC). */
  readonly redoubleMin: number;
}

// ─── Suit response parameters ───────────────────────────────

export type SuitResponseForcingDuration = "one-round" | "game";
export type OneNtForcingStatus = "non-forcing" | "forcing" | "semi-forcing";

export interface SuitResponseConfig {
  /** Minimum HCP for a 2-level new-suit response to a 1-level suit opening.
   *  SAYC: 10 (forcing one round). 2/1: 12 (game-forcing). */
  readonly twoLevelMin: number;
  /** Forcing duration of a 2-level new-suit response.
   *  SAYC: "one-round". 2/1: "game". */
  readonly twoLevelForcingDuration: SuitResponseForcingDuration;
}

export interface OneNtResponseAfterMajorConfig {
  /** Whether 1NT response to 1M is forcing.
   *  SAYC: "non-forcing". 2/1: "semi-forcing" (opener may pass 5332 min).
   *  Strict 2/1: "forcing". */
  readonly forcing: OneNtForcingStatus;
  /** Maximum HCP for 1NT response to 1M.
   *  SAYC: 10. 2/1: 12. */
  readonly maxHcp: number;
  /** Minimum HCP for 1NT response to 1M (e.g. 6 in SAYC/Acol). */
  readonly minHcp: number;
}

// ─── DONT overcall parameters ────────────────────────────────

/** HCP bounds for DONT overcalls — may vary by system. */
export interface DontOvercallConfig {
  /** Minimum HCP for a DONT overcall (e.g. 8). */
  readonly minHcp: number;
  /** Maximum HCP for a DONT overcall (e.g. 15). */
  readonly maxHcp: number;
}

// ─── Opening requirements ───────────────────────────────────

/** Opening bid structural requirements that vary by system. */
export interface OpeningRequirements {
  /** Minimum suit length to open a major (4 in Acol, 5 in SAYC/2/1). */
  readonly majorSuitMinLength: 4 | 5;
}

// ─── Top-level SystemConfig ─────────────────────────────────

export interface SystemConfig {
  /** System identifier — must be a value from base-system-vocabulary.ts. */
  readonly systemId: BaseSystemId;
  /** Human-readable name (e.g. "Standard American Yellow Card"). */
  readonly displayName: string;

  /** 1NT opening HCP range. */
  readonly ntOpening: NtOpeningConfig;
  /** Responder point-range thresholds opposite a 1NT opening. */
  readonly responderThresholds: ResponderThresholds;
  /** Opener rebid thresholds (e.g. super-accept, not-minimum). */
  readonly openerRebid: OpenerRebidThresholds;
  /** Interference thresholds. */
  readonly interference: InterferenceThresholds;
  /** 2-level new-suit response parameters (forcing level, HCP threshold). */
  readonly suitResponse: SuitResponseConfig;
  /** 1NT response to 1M parameters (forcing status, HCP ceiling). */
  readonly oneNtResponseAfterMajor: OneNtResponseAfterMajorConfig;
  /** Structural opening bid requirements (major length, etc.). */
  readonly openingRequirements: OpeningRequirements;
  /** DONT overcall HCP bounds. */
  readonly dontOvercall: DontOvercallConfig;
}

// ─── Concrete system configs ────────────────────────────────

export const SAYC_SYSTEM_CONFIG: SystemConfig = {
  systemId: BASE_SYSTEM_SAYC,
  displayName: "Standard American Yellow Card",
  ntOpening: { minHcp: 15, maxHcp: 17 },
  responderThresholds: {
    inviteMin: 8,
    inviteMax: 9,
    gameMin: 10,
    slamMin: 15,
  },
  openerRebid: { notMinimum: 16 },
  interference: { redoubleMin: 10 },
  suitResponse: { twoLevelMin: 10, twoLevelForcingDuration: "one-round" },
  oneNtResponseAfterMajor: { forcing: "non-forcing", maxHcp: 10, minHcp: 6 },
  openingRequirements: { majorSuitMinLength: 5 },
  dontOvercall: { minHcp: 8, maxHcp: 15 },
};

export const TWO_OVER_ONE_SYSTEM_CONFIG: SystemConfig = {
  systemId: BASE_SYSTEM_TWO_OVER_ONE,
  displayName: "2/1 Game Forcing",
  ntOpening: { minHcp: 15, maxHcp: 17 },
  responderThresholds: {
    inviteMin: 8,
    inviteMax: 9,
    gameMin: 10,
    slamMin: 15,
  },
  openerRebid: { notMinimum: 16 },
  interference: { redoubleMin: 10 },
  suitResponse: { twoLevelMin: 12, twoLevelForcingDuration: "game" },
  oneNtResponseAfterMajor: { forcing: "semi-forcing", maxHcp: 12, minHcp: 6 },
  openingRequirements: { majorSuitMinLength: 5 },
  dontOvercall: { minHcp: 8, maxHcp: 15 },
};

export const ACOL_SYSTEM_CONFIG: SystemConfig = {
  systemId: BASE_SYSTEM_ACOL,
  displayName: "Acol",
  // Standard weak NT: 12-14 HCP (EBU standard, UK tournament play)
  ntOpening: { minHcp: 12, maxHcp: 14 },
  responderThresholds: {
    // Thresholds shifted to keep 25 combined for game (EBU standard Acol)
    inviteMin: 10,  // 10+12=22 combined; invite with 22-24, game at 25+
    inviteMax: 12,  // invite range: 10-12
    gameMin: 13,    // 13+12=25 combined for game
    slamMin: 19,    // 19+14=33 combined for slam
  },
  // minHcp+1 pattern: accepts invite with 13-14, declines with 12
  openerRebid: { notMinimum: 13 },
  // 9+12=21 min combined for penalty interest (cf. SAYC: 10+15=25). Assumes strength redouble — not Wriggle/SOS escape.
  interference: { redoubleMin: 9 },
  // Practical guideline; Acol forces new suit by structure regardless of HCP. Simplified for teaching.
  suitResponse: { twoLevelMin: 10, twoLevelForcingDuration: "one-round" },
  // "Dustbin 1NT": non-forcing, capped at 9; with 10+ responder bids a new suit
  oneNtResponseAfterMajor: { forcing: "non-forcing", maxHcp: 9, minHcp: 6 },
  openingRequirements: { majorSuitMinLength: 4 },
  dontOvercall: { minHcp: 8, maxHcp: 15 },
};

// ─── System config registry ─────────────────────────────────

/** UI-friendly metadata for each base system. */
export interface BaseSystemMeta {
  readonly id: BaseSystemId;
  readonly label: string;
  readonly shortLabel: string;
}

/** Available base systems for the UI selector (order = display order). */
export const AVAILABLE_BASE_SYSTEMS: readonly BaseSystemMeta[] = [
  { id: BASE_SYSTEM_SAYC, label: "Standard American Yellow Card", shortLabel: "SAYC" },
  { id: BASE_SYSTEM_TWO_OVER_ONE, label: "2/1 Game Forcing", shortLabel: "2/1" },
  { id: BASE_SYSTEM_ACOL, label: "Acol", shortLabel: "Acol" },
] as const;

const SYSTEM_CONFIG_MAP: Readonly<Record<string, SystemConfig>> = {
  [BASE_SYSTEM_SAYC]: SAYC_SYSTEM_CONFIG,
  [BASE_SYSTEM_TWO_OVER_ONE]: TWO_OVER_ONE_SYSTEM_CONFIG,
  [BASE_SYSTEM_ACOL]: ACOL_SYSTEM_CONFIG,
};

/** Look up SystemConfig by base system id. Falls back to SAYC for unknown ids. */
export function getSystemConfig(id: BaseSystemId): SystemConfig {
  return SYSTEM_CONFIG_MAP[id] ?? SAYC_SYSTEM_CONFIG;
}

File: src-tauri/crates/bridge-conventions/src/types/system_config.rs
//! Base bidding system configuration.
//!
//! Mirrors TS types from `conventions/definitions/system-config.ts`.

use serde::{Deserialize, Serialize};

/// Point formula identifiers for composing total-point values.
///
/// The engine computes raw components (HCP, shortage, length).
/// The fact DSL composes them using these formula IDs via `compute_total_points()`.
#[derive(Debug, Clone, Copy, PartialEq, Eq, Hash, Serialize, Deserialize)]
#[serde(rename_all = "kebab-case")]
pub enum PointFormulaId {
    HcpOnly,
    HcpPlusShortage,
    HcpPlusAllDistribution,
}

/// Point formula configuration per contract type (NT vs trump).
#[derive(Debug, Clone, Copy, PartialEq, Eq, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct PointConfig {
    pub nt_formula: PointFormulaId,
    pub trump_formula: PointFormulaId,
}

fn default_point_config() -> PointConfig {
    PointConfig {
        nt_formula: PointFormulaId::HcpOnly,
        trump_formula: PointFormulaId::HcpPlusShortage,
    }
}

fn is_default_point_config(config: &PointConfig) -> bool {
    *config == default_point_config()
}

/// Base bidding system identifiers.
#[derive(Debug, Clone, Copy, PartialEq, Eq, Hash, Serialize, Deserialize)]
pub enum BaseSystemId {
    #[serde(rename = "sayc")]
    Sayc,
    #[serde(rename = "two-over-one")]
    TwoOverOne,
    #[serde(rename = "acol")]
    Acol,
}

/// Suit total-point equivalents for a threshold.
#[derive(Debug, Clone, Copy, PartialEq, Eq, Serialize, Deserialize)]
pub struct TotalPointEquivalent {
    pub trump: u32,
}

/// 1NT opening HCP range.
#[derive(Debug, Clone, Copy, PartialEq, Eq, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct NtOpeningConfig {
    pub min_hcp: u32,
    pub max_hcp: u32,
}

/// Responder point-range thresholds opposite a 1NT opening.
#[derive(Debug, Clone, Copy, PartialEq, Eq, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct ResponderThresholds {
    pub invite_min: u32,
    pub invite_max: u32,
    pub game_min: u32,
    pub slam_min: u32,
    pub invite_min_tp: TotalPointEquivalent,
    pub invite_max_tp: TotalPointEquivalent,
    pub game_min_tp: TotalPointEquivalent,
    pub slam_min_tp: TotalPointEquivalent,
}

/// Opener rebid thresholds.
#[derive(Debug, Clone, Copy, PartialEq, Eq, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct OpenerRebidThresholds {
    pub not_minimum: u32,
    pub not_minimum_tp: TotalPointEquivalent,
}

/// Interference thresholds.
#[derive(Debug, Clone, Copy, PartialEq, Eq, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct InterferenceThresholds {
    pub redouble_min: u32,
}

/// Forcing duration of a 2-level new-suit response.
#[derive(Debug, Clone, Copy, PartialEq, Eq, Hash, Serialize, Deserialize)]
#[serde(rename_all = "kebab-case")]
pub enum SuitResponseForcingDuration {
    OneRound,
    Game,
}

/// Whether 1NT response to 1M is forcing.
#[derive(Debug, Clone, Copy, PartialEq, Eq, Hash, Serialize, Deserialize)]
#[serde(rename_all = "kebab-case")]
pub enum OneNtForcingStatus {
    NonForcing,
    Forcing,
    SemiForcing,
}

/// 2-level new-suit response parameters.
#[derive(Debug, Clone, Copy, PartialEq, Eq, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct SuitResponseConfig {
    pub two_level_min: u32,
    pub two_level_forcing_duration: SuitResponseForcingDuration,
}

/// 1NT response to 1M parameters.
#[derive(Debug, Clone, Copy, PartialEq, Eq, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct OneNtResponseAfterMajorConfig {
    pub forcing: OneNtForcingStatus,
    pub max_hcp: u32,
    pub min_hcp: u32,
}

/// DONT overcall HCP bounds.
#[derive(Debug, Clone, Copy, PartialEq, Eq, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct DontOvercallConfig {
    pub min_hcp: u32,
    pub max_hcp: u32,
}

/// Structural opening bid requirements.
#[derive(Debug, Clone, Copy, PartialEq, Eq, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct OpeningRequirements {
    pub major_suit_min_length: u8,
}

/// Top-level system configuration.
#[derive(Debug, Clone, PartialEq, Eq, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct SystemConfig {
    pub system_id: BaseSystemId,
    pub display_name: String,
    pub nt_opening: NtOpeningConfig,
    pub responder_thresholds: ResponderThresholds,
    pub opener_rebid: OpenerRebidThresholds,
    pub interference: InterferenceThresholds,
    pub suit_response: SuitResponseConfig,
    pub one_nt_response_after_major: OneNtResponseAfterMajorConfig,
    pub opening_requirements: OpeningRequirements,
    pub dont_overcall: DontOvercallConfig,
    #[serde(default = "default_point_config", skip_serializing_if = "is_default_point_config")]
    pub point_config: PointConfig,
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn base_system_id_serde() {
        assert_eq!(
            serde_json::to_string(&BaseSystemId::Sayc).unwrap(),
            "\"sayc\""
        );
        assert_eq!(
            serde_json::to_string(&BaseSystemId::TwoOverOne).unwrap(),
            "\"two-over-one\""
        );
        assert_eq!(
            serde_json::to_string(&BaseSystemId::Acol).unwrap(),
            "\"acol\""
        );
    }

    #[test]
    fn forcing_duration_serde() {
        assert_eq!(
            serde_json::to_string(&SuitResponseForcingDuration::OneRound).unwrap(),
            "\"one-round\""
        );
        assert_eq!(
            serde_json::to_string(&SuitResponseForcingDuration::Game).unwrap(),
            "\"game\""
        );
    }

    #[test]
    fn one_nt_forcing_status_serde() {
        assert_eq!(
            serde_json::to_string(&OneNtForcingStatus::NonForcing).unwrap(),
            "\"non-forcing\""
        );
        assert_eq!(
            serde_json::to_string(&OneNtForcingStatus::SemiForcing).unwrap(),
            "\"semi-forcing\""
        );
    }

    #[test]
    fn system_config_roundtrip() {
        let config = SystemConfig {
            system_id: BaseSystemId::Sayc,
            display_name: "Standard American Yellow Card".to_string(),
            nt_opening: NtOpeningConfig {
                min_hcp: 15,
                max_hcp: 17,
            },
            responder_thresholds: ResponderThresholds {
                invite_min: 8,
                invite_max: 9,
                game_min: 10,
                slam_min: 15,
                invite_min_tp: TotalPointEquivalent { trump: 8 },
                invite_max_tp: TotalPointEquivalent { trump: 10 },
                game_min_tp: TotalPointEquivalent { trump: 10 },
                slam_min_tp: TotalPointEquivalent { trump: 16 },
            },
            opener_rebid: OpenerRebidThresholds {
                not_minimum: 16,
                not_minimum_tp: TotalPointEquivalent { trump: 16 },
            },
            interference: InterferenceThresholds { redouble_min: 10 },
            suit_response: SuitResponseConfig {
                two_level_min: 10,
                two_level_forcing_duration: SuitResponseForcingDuration::OneRound,
            },
            one_nt_response_after_major: OneNtResponseAfterMajorConfig {
                forcing: OneNtForcingStatus::NonForcing,
                max_hcp: 10,
                min_hcp: 6,
            },
            opening_requirements: OpeningRequirements {
                major_suit_min_length: 5,
            },
            dont_overcall: DontOvercallConfig {
                min_hcp: 8,
                max_hcp: 15,
            },
            point_config: PointConfig {
                nt_formula: PointFormulaId::HcpOnly,
                trump_formula: PointFormulaId::HcpPlusShortage,
            },
        };
        let json = serde_json::to_string(&config).unwrap();
        let back: SystemConfig = serde_json::from_str(&json).unwrap();
        assert_eq!(back, config);
    }

    #[test]
    fn point_formula_id_serde() {
        assert_eq!(
            serde_json::to_string(&PointFormulaId::HcpOnly).unwrap(),
            "\"hcp-only\""
        );
        assert_eq!(
            serde_json::to_string(&PointFormulaId::HcpPlusShortage).unwrap(),
            "\"hcp-plus-shortage\""
        );
        assert_eq!(
            serde_json::to_string(&PointFormulaId::HcpPlusAllDistribution).unwrap(),
            "\"hcp-plus-all-distribution\""
        );
    }

    #[test]
    fn point_config_default_on_missing_field() {
        // Ensure SystemConfig without pointConfig deserializes with default
        let json = r#"{"systemId":"sayc","displayName":"test","ntOpening":{"minHcp":15,"maxHcp":17},"responderThresholds":{"inviteMin":8,"inviteMax":9,"gameMin":10,"slamMin":15,"inviteMinTp":{"trump":8},"inviteMaxTp":{"trump":10},"gameMinTp":{"trump":10},"slamMinTp":{"trump":16}},"openerRebid":{"notMinimum":16,"notMinimumTp":{"trump":16}},"interference":{"redoubleMin":10},"suitResponse":{"twoLevelMin":10,"twoLevelForcingDuration":"one-round"},"oneNtResponseAfterMajor":{"forcing":"non-forcing","maxHcp":10,"minHcp":6},"openingRequirements":{"majorSuitMinLength":5},"dontOvercall":{"minHcp":8,"maxHcp":15}}"#;
        let config: SystemConfig = serde_json::from_str(json).unwrap();
        assert_eq!(config.point_config.nt_formula, PointFormulaId::HcpOnly);
        assert_eq!(config.point_config.trump_formula, PointFormulaId::HcpPlusShortage);
    }
}
